marketplace/src/lib.rs

```diff
@@ -83,679 +76,503 @@ pub mod pallet {
+
+				// Execute
+				config_op_field_exp!(marketplace.commission_fee, commission_fee);
+				config_op_field_exp!(marketplace.listing_fee, listing_fee);
+				config_op_field_exp!(marketplace.account_list, account_list.clone());
+				config_op_field_exp!(marketplace.offchain_data, offchain_data.clone());
 				Ok(())
 			})?;
 
-			let event = Event::MarketplaceNameUpdated { marketplace_id, name };
+			let event = Event::MarketplaceConfigSet {
+				marketplace_id,
+				commission_fee,
+				listing_fee,
+				account_list,
+				offchain_data,
+			};
 			Self::deposit_event(event);
 
 			Ok(().into())
 		}
 
+		/// Set the fee for minting a marketplace if the caller is root.
 		#[pallet::weight(T::WeightInfo::set_marketplace_mint_fee())]
 		pub fn set_marketplace_mint_fee(
 			origin: OriginFor<T>,
-			mint_fee: BalanceOf<T>,
+			fee: BalanceOf<T>,
 		) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
-
-			MarketplaceMintFee::<T>::put(mint_fee);
-
-			Self::deposit_event(Event::MarketplaceMintFeeUpdated { fee: mint_fee });
+			MarketplaceMintFee::<T>::put(fee);
+			Self::deposit_event(Event::MarketplaceMintFeeSet { fee });
 
 			Ok(().into())
 		}
 
-		#[pallet::weight(T::WeightInfo::set_marketplace_commission_fee())]
-		pub fn set_marketplace_commission_fee(
+		/// Put an NFT on sale on a marketplace.
+		#[pallet::weight(T::WeightInfo::list_nft())]
+		#[transactional]
+		pub fn list_nft(
 			origin: OriginFor<T>,
+			nft_id: NFTId,
+			price: BalanceOf<T>,
 			marketplace_id: MarketplaceId,
-			commission_fee: u8,
 		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			ensure!(commission_fee <= 100, Error::<T>::InvalidCommissionFeeValue);
 
-			Marketplaces::<T>::mutate(marketplace_id, |x| -> DispatchResult {
-				let market_info = x.as_mut().ok_or(Error::<T>::MarketplaceNotFound)?;
-				ensure!(market_info.owner == who, Error::<T>::NotMarketplaceOwner);
-				market_info.commission_fee = commission_fee;
-				Ok(())
-			})?;
+			// Checks
+			let nft = T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFound)?;
+			ensure!(nft.owner == who, Error::<T>::NotTheNFTOwner);
+			ensure!(!nft.state.listed_for_sale, Error::<T>::NFTAlreadyListed);
+			ensure!(!nft.state.is_capsule, Error::<T>::CannotListCapsuleNFTs);
+			ensure!(!nft.state.is_delegated, Error::<T>::CannotListDelegatedNFTs);
+			ensure!(!nft.state.is_soulbound, Error::<T>::CannotListSoulboundNFTs);
+
+			let marketplace =
+				Marketplaces::<T>::get(marketplace_id).ok_or(Error::<T>::MarketplaceNotFound)?;
+
+			// Check if the user is allowed to list on this marketplace.
+			Self::ensure_is_allowed_to_list(&who, &marketplace)?;
+
+			// Check if the selected price can cover the marketplace commission_fee if it exists.
+			if let Some(commission_fee) = &marketplace.commission_fee {
+				if let MarketplaceFee::Flat(flat_commission) = commission_fee {
+					ensure!(price >= *flat_commission, Error::<T>::PriceTooLowForCommissionFee);
```

Comment:
Maybe change the error name to "PriceCannotCoverMarketplaceFee"? something like that